Switch Operator Processors

matthew-reed-holden opened the pull request:
<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one or more
  ~ contributor license agreements.  See the NOTICE file distributed with
  ~ this work for additional information regarding copyright ownership.
  ~ The ASF licenses this file to You under the Apache License, Version 2.0
  ~ (the "License"); you may not use this file except in compliance with
  ~ the License.  You may obtain a copy of the License at
  ~
  ~    http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  ~
  -->

  <!--
Thanks for contributing! Here are some tips you can follow to help us incorporate your contribution quickly and easily:
1. If this is your first time, please read our contributor guidelines:
    - https://streampipes.apache.org/community/get-involved/
    - https://cwiki.apache.org/confluence/display/STREAMPIPES/Getting+Started
2. Make sure the PR title is formatted like: `[#<GitHub issue id>] PR title ...`
3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., `[WIP][#<GitHub issue id>] PR title ...`.
4. Please write your PR title to summarize what this PR proposes/fixes.
5. Link the PR to the corresponding GitHub issue (if present) in the `Development` section in the right menu bar.
6. Be sure to keep the PR description updated to reflect all changes.
7. If possible, provide a concise example to reproduce the issue for a faster review.
8. Make sure tests pass via `mvn clean install`.
9. (Optional) If the contribution is large, please file an Apache ICLA
    - http://apache.org/licenses/icla.pdf
-->

### Purpose
<!--
Please clarify what changes you are proposing and describe how those changes will address the issue.
Furthermore, describe potential consequences the changes might have.
-->
Introduce Pipeline Element Processors that allow a user to implement the "Switch" operator. I have often found myself wanting to setup a pipeline that uses the "Switch" operator but have had to use a complicated set of boolean conditions to try and implement it or I eventually end up just using a JavaScript processor to use the switch statement. The processors reflect the following block.

```java
switch(String x) {
  case "test":
   do something;
   break;
  case "test 2":
   do something else;
   break;
  default:
   do this if nothing else works;
   break;
}
```

The proposed updates introduce three processors, each which have the purpose of handling a different input type. I have implemented a Boolean Input processor, String input processor, and Numerical input processor. This is done primarily for the sake of cleaner code along with simplifying the detection of the incoming event property. These three processors could be combined into one "Switch Operator Processor" if there was the introduction of some "Conditional Requirements" component.

![image](https://github.com/user-attachments/assets/161d9c80-7848-49c8-b6d4-7fe003b70781)

![image](https://github.com/user-attachments/assets/5fe43290-14fd-4750-a680-9a3eae494ba6)

![image](https://github.com/user-attachments/assets/bd256fa0-c162-4ab5-9b68-da5557f8e7be)

![image](https://github.com/user-attachments/assets/c7478af1-e6fa-4e10-a174-c10c7ffd6db5)

### Remarks
<!--
Is there anything left we need to pay attention on?
Are there some references that might be important? E.g. links to Confluence, or discussions
on the mailing list or GitHub.
-->

I have written unit tests for the numerical processor locally but ran into issues when I transferred the updates from my custom extensions repo to the main streampipes codebase. If someone has input on the right way to write tests for streampipes or would like to write the unit tests themselves that would be welcome.

PR introduces (a) breaking change(s): no
PR introduces (a) deprecation(s): no

matthew-reed-holden committed:
feat: Created processor to perform the "switch" operator in streampipes pipelines.

- define "Switch" input. e.g. switch(event.switch_input)
- define "cases" which output string types
- define "default" case which evaluates if all other cases fail
- creates "switch-filter-result" output
- add SwitchOperatorProcessor to TransformationExtensionModuleExport

matthew-reed-holden committed:
feat: Add documentation and localization for Switch Operator processor
- added documentation.md based on generated template
- added strings.en resources
- added new icon of switch to indicate function of processor

chore: Rename the processor to use the correct format of org.apache.org...

matthew-reed-holden committed:
chore: Unit testing for SwitchOperatorProcessorTest

- Need to fix test issues of Mockito not being able to mock

matthew-reed-holden committed:
test: Refactor and enhance unit tests for SwitchOperatorProcessor

matthew-reed-holden committed:
feat: Enhance SwitchOperatorProcessor to support multiple output types

- Update to return primitive data types based on user selection
- Updated tests to cover functionality

matthew-reed-holden committed:
feat: Update documentation for Switch Operator processor to include output type options and examples

matthew-reed-holden committed:
feat: Implement LogicalOperator and SwitchBooleanInputProcessor with enhanced output type handling

matthew-reed-holden committed:
feat: Add SwitchNumericalInputProcessor and SwitchStringInputProcessor for enhanced switch functionality

matthew-reed-holden committed:
feat: Add Switch Operator documentation and refactor processor interfaces for improved clarity

matthew-reed-holden committed:
feat: Add license headers to switch operator processor files for compliance

matthew-reed-holden committed:
feat: Remove unused constants from Switch input processors for cleaner code

matthew-reed-holden committed:
Merge branch 'apache:dev' into switch-operator-processor

tenthe commented:
Hi @heisenbergs-uncertainty,

Thanks a lot for the PR — these are great new processing elements, and your description makes it easy to review.

* For unit tests, you can use the `ProcessingElementTestExecutor` helper class.

* See `TransformToBooleanProcessorTest` or `NumericalFilterProcessorTest` for examples.

Would you like to add some unit tests for your elements?
Feel free to reach out if you have any questions or suggestions for improving the test framework.

Hope this helps!

matthew-reed-holden committed:
feat: Introduce AbstractSwitchOperatorProcessor and refactor switch input processors for improved structure and maintainability

- Refactored to use AbstractSwitchOperatorProcessor to pull out common methods
- Overall Renaming of methods and names for better consistency and descriptions of functionality
- moved general implementation of onEvent() to AbstractSwitchOperatorProcessor which utilizes findMatchingResult which is then given class specific implementations

matthew-reed-holden committed:
feat: Refactor switch operator processors to use common constants and add unit tests for boolean, numerical, and string input processors

matthew-reed-holden committed:
feat: Add support for boolean, numerical, and string inputs in Switch Operator with updated documentation and license headers
- Updated documentation.md and strings.en to reflect the changes made in optimization refactor
- Unit tests added to cover general functionality and edge cases
- Utilized ProcessingElementTestExecutor for the unit tests
- Added Licenses to top of Test files and new optimized classes

matthew-reed-holden committed:
chore: checkstyle updates

matthew-reed-holden committed:
docs: Add README for StreamPipes Switch Operators with detailed overview and usage instructions

matthew-reed-holden commented:
> Hi @heisenbergs-uncertainty,
> 
> Thanks a lot for the PR — these are great new processing elements, and your description makes it easy to review.
> 
> * For unit tests, you can use the `ProcessingElementTestExecutor` helper class.
> * See `TransformToBooleanProcessorTest` or `NumericalFilterProcessorTest` for examples.
> 
> Would you like to add some unit tests for your elements? Feel free to reach out if you have any questions or suggestions for improving the test framework.
> 
> Hope this helps!

Hey @tenthe! Thank you for the testing recommendation, that was super useful and made writing the tests much simpler. Props to whoever wrote that suite!

I have gone ahead and added unit tests for the processors and made quite a few updates that they helped me catch. From my testing it should be good to go at this point.

tenthe commented on streampipes-extensions/streampipes-processors-transformation-jvm/src/main/java/org/apache/streampipes/processors/transformation/jvm/processor/switchoperator/AbstractSwitchOperatorProcessor.java:
```diff
@@ -0,0 +1,148 @@
+   * @param staticPropertyExtractor Extractor for the current switch case group.
+   * @return A SwitchCaseEntry object.
+   */
+  protected SwitchCaseEntry parseSwitchCaseEntry(StaticPropertyExtractor staticPropertyExtractor) {
```
Can this be an abstract method?

matthew-reed-holden replied:
I don't think so. Switching it to abstract is causing the classes that extend it to refuse to implement the class.

But I am still refreshing my understanding of Java so I will default to your decision.

tenthe replied:
Thanks for your response. This was just a small point I noticed while reviewing the code. I'll trigger the GitHub Action validation now. Once it passes, we can proceed with merging your PR.

tenthe reviewed (approved):
Hi, thanks for the changes.
I just had a minor comment, but I think the code looks good from my side.

I look forward to use the processors.
Cheers,
Philipp

matthew-reed-holden committed:
chore: Add missing license headers in processor classes and README
